### PR TITLE
feat: deploy Nutanix CCM as an Addon

### PR DESCRIPTION
**Pull request by dkoshkin**
Stacked on [dkoshkin/fix-nutanix-csi](https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/34)

This PR removes the Nutanix CCM from the templates and instead deploys it using CAAPH using the published Helm chart.

```
$ kubectl get pods -n kube-system
NAME                                                          READY   STATUS    RESTARTS   AGE
coredns-5dd5756b68-d8t7r                                      0/1     Running   0          83s
coredns-5dd5756b68-z6l4g                                      0/1     Running   0          83s
etcd-dkoshkin-nutanix-0508-2-w9kmm-7mkl2                      1/1     Running   0          95s
kube-apiserver-dkoshkin-nutanix-0508-2-w9kmm-7mkl2            1/1     Running   0          95s
kube-controller-manager-dkoshkin-nutanix-0508-2-w9kmm-7mkl2   1/1     Running   0          95s
kube-proxy-7x762                                              1/1     Running   0          52s
kube-proxy-hg5mq                                              1/1     Running   0          83s
kube-scheduler-dkoshkin-nutanix-0508-2-w9kmm-7mkl2            1/1     Running   0          96s
kube-vip-dkoshkin-nutanix-0508-2-w9kmm-7mkl2                  1/1     Running   0          95s
nutanix-cloud-controller-manager-694ff9d9dd-m9z92             1/1     Running   0          83s


$ kubectl get nodes -o yaml | grep providerID
    providerID: nutanix://11e638da-6aa0-45b9-8477-3836aa965842
    providerID: nutanix://db91a5ca-22a0-4818-8b5e-c79361f82877
```

https://d2iq.atlassian.net/browse/D2IQ-99841

### CODE REVIEW

**Review by faiq — APPROVED** [review]
looks good

File: api/v1alpha1/addon_types.go [hunk 1]
```diff
@@ -308,12 +308,34 @@ func (CSI) VariableSchema() clusterv1.VariableSchema {
...
+type CCM struct {
+	// A reference to the Secret for credential information for the target Prism Central instance
+	// +optional
+	Credentials *corev1.LocalObjectReference `json:"credentials"`
+}
 
 func (CCM) VariableSchema() clusterv1.VariableSchema {
+	// TODO Validate credentials is set.
+	// This CCM is shared across all providers.
+	// Some of these providers may require credentials to be set, but we don't want to require it for all providers.
```
Comment (supershal) [inline api/v1alpha1/addon_types.go:320]: great details

File: pkg/handlers/generic/lifecycle/ccm/handler.go [hunk 2]
```diff
@@ -88,17 +88,39 @@ func (c *CCMHandler) AfterControlPlaneInitialized(
...
+		clusterconfig.MetaVariableName,
+	)
+	if err != nil {
+		log.Error(
+			err,
+			"failed to read clusterConfig variable from cluster definition",
+		)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(
+			fmt.Sprintf("ffailed to read clusterConfig variable from cluster definition: %v",
```
Comment (supershal) [inline pkg/handlers/generic/lifecycle/ccm/handler.go:103]:
```suggestion
			fmt.Sprintf("failed to read clusterConfig variable from cluster definition: %v",
```

File: charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2024 D2iQ, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+{{- if .Values.hooks.clusterAutoscaler.helmAddonStrategy.defaultValueTemplateConfigMap.create }}
```
Comment (supershal) [inline charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml:4]:
```suggestion
{{- if .Values.hooks.ccm.helmAddonStrategy.defaultValueTemplateConfigMap.create }}
```

File: charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+kind: ConfigMap
+metadata:
+  name: '{{ .Values.hooks.ccm.nutanix.helmAddonStrategy.defaultValueTemplateConfigMap.name }}'
+data:
+  values.yaml: |-
+    ---
+    prismCentralEndPoint: {{ `{{ .PrismCentralHost }}` }}
+    prismCentralPort: {{ `{{ .PrismCentralPort }}` }}
+    prismCentralInsecure: {{ `{{ .PrismCentralInsecure }}` }}
+    prismCentralAdditionalTrustBundle: {{ `"{{ or .PrismCentralAdditionalTrustBundle "" }}"`  }}
```
Comment (supershal) [inline charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml:15]: Do we have to specify Prism endpoint related values here?
Can these values template be in the Handler code?
Handler can merge data from `values.yaml` and the parsed template for prism endpoint while creating helm proxy spec.
i.e. keep both the template and parsing logic in the handler code.

File: pkg/handlers/generic/lifecycle/ccm/nutanix/handler.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	helmChartInfoGetter *config.HelmChartGetter,
+) *provider {
+	return &provider{
+		client:              c,
+		config:              cfg,
+		helmChartInfoGetter: helmChartInfoGetter,
+	}
+}
+
+func (p *provider) Apply(
```
Comment (faiq) [inline pkg/handlers/generic/lifecycle/ccm/nutanix/handler.go:72]: Do we need to look at the strategy at all here?

File: hack/addons/update-nutanix-ccm.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+set -euo pipefail
```
Comment (faiq) [inline hack/addons/update-nutanix-ccm.sh:2]: I thought about these again do we really need this file here if we're not going to support this with a CRS strategy?

Comment (dkoshkin) [reply hack/addons/update-nutanix-ccm.sh]: I added this originally thinking we can use this to capture the container images that will be needed, but we don't have that process yet, so removing it https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/38/commits/abde55db57ac35e9465d2abbf37326ebd504289a

Thanks for the suggestion.

File: charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml (hunk 3 above)
Comment (dkoshkin) [reply charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml]: Fixed in https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/38/commits/b7f79b8fc3c51c1789900a4bf7be42fe5a9b1ccd

File: pkg/handlers/generic/lifecycle/ccm/handler.go (hunk 2 above)
Comment (dkoshkin) [reply pkg/handlers/generic/lifecycle/ccm/handler.go]: Fixed in https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/38/commits/cd8dd6d7701a86152c195d0ac150d2ef705ad177

File: pkg/handlers/generic/lifecycle/ccm/nutanix/handler.go (hunk 5 above)
Comment (dkoshkin) [reply pkg/handlers/generic/lifecycle/ccm/nutanix/handler.go:72]: No the CCM API does not expose strategy, lets not add it until its absolutely necessary.

File: charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml (hunk 4 above)
Comment (supershal) [reply charts/cluster-api-runtime-extensions-nutanix/templates/ccm/nutanix/manifests/helm-addon-installation.yaml:15]: Thank you Dimitri.

**Review by supershal — APPROVED** [review]
LGTM.

**Pull request merged**